pkg/frontend/querymiddleware/durations.go

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+// newDurationsMiddleware returns a MetricsQueryMiddleware that evaluates and
+// rewrites duration expressions in the query.
+// As long as durations can be calculated solely from consts and the step
+// this works. When that is not the case anymore, we'll have to implement
+// checks before sharding/splitting/etc middlewares individually to check what
+// we _can_ support.
+func newDurationsMiddleware(
+	logger log.Logger,
+) MetricsQueryMiddleware {
+	return MetricsQueryMiddlewareFunc(func(next MetricsQueryHandler) MetricsQueryHandler {
+		return &durationsMiddleware{
+			next:   next,
+			logger: logger,
+		}
+	})
+}
+
+type durationsMiddleware struct {
+	next   MetricsQueryHandler
+	logger log.Logger
+}
+
+func (d *durationsMiddleware) Do(ctx context.Context, req MetricsQueryRequest) (Response, error) {
+	var err error
+	req, err = d.rewriteIfNeeded(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return d.next.Do(ctx, req)
+}
+
+func (d *durationsMiddleware) rewriteIfNeeded(ctx context.Context, req MetricsQueryRequest) (MetricsQueryRequest, error) {
+	// Log the instant query and its timestamp in every error log, so that we have more information for debugging failures.
+	logger := log.With(d.logger, "query", req.GetQuery(), "query_timestamp", req.GetStart())
+
+	spanLog, _ := spanlogger.New(ctx, logger, tracer, "durationsMiddleware.Do")
+	defer spanLog.Finish()
+
+	origQuery := req.GetQuery()
+	expr, err := parser.ParseExpr(origQuery)
+	if err != nil {
+		// This middleware focuses on duration expressions, so if the query is
+		// not valid, we just fall through to the next handler.
+		return req, nil
+	}
+
+	expr, err = promql.PreprocessExpr(expr, time.UnixMilli(req.GetStart()), time.UnixMilli(req.GetEnd()), time.Duration(req.GetStep())*time.Second)
```

Comment:
`PreprocessExpr` will also resolve the value of timestamps in expressions like `foo @ start()`. We already have logic to replace `start()` and `end()` with the actual query timestamp in the query splitting logic (`evaluateAtModifierFunction` in `split_and_cache.go`), so now we're duplicating work.

Using `PreprocessExpr` to do this seems like the most robust way to do this - should we remove this from the split and cache middleware?

Reply:
I didn't actually catch that. I've added tests for `start()` and `end()`.
It probably makes sense to remove from the split and cache, but it's up to you. Do you want me to do it in this PR?

Reply:
>Do you want me to do it in this PR?

I'm happy for it to be a follow-up PR.